Added configurable interpolation resolution scaling, additional information in README, and smartprobe.compensation.ngc to produce probe-results.txt

DauntlessAq opened the pull request:
The primary purpose of this pull request is to merge configurable interpolation resolution into compensation.py
Additionally, a probing program, _smartprobe_compensation.ngc_, is included, and the README has increased in detail and documents the new features.

This pull request includes no additional dependencies.

**Changes to compensation.py:**
_Added ‘resolution’ HAL pin for specifying the interpolation resolution._
The value of resolution defaults to ‘1’ when initiated, giving an interpolation resolution of 1mm. This is the same as the original behaviour.
However, increasing the resolution allows for use of compensation on parts with arbitrary roughness where 1mm interpolation is insufficiently dense to capture the surface, and to benefit from higher resolution surface maps.
Changing the resolution takes effect when the compensation map is recalculated.

**New file – smartprobe_compensation.ngc:**
This file is [smartprobe.ngc]( https://github.com/LinuxCNC/linuxcnc/blob/master/nc_files/smartprobe.ngc) with some minor adaptions to work with compensation.py, primarily changing the log to format ‘x y z ‘ data correctly and name the file ‘probe-results.txt’
Further modifications made are not essential for operation with compensation.py, but they add some additional functionality:
Switched to metric from imperial. Changed to named global variables, added annotations. Added optional slow probe move. Added 'walking' logic to Y moves to avoid retract to Z safety to wastes time. Added velocity control to most move types. Changed to drive the program by final dimensions, with option to round number of rows and/or columns either up or down if the result is not an integer.

**Changes to README:**
_Added recommendation in README to connect compensation.x-pos and compensation.y-pos to halui.axis.x.pos-relative and halui.axis.y.pos-relative respectively:_
This change means that the compensation map is applied in the active work offsets of the active work coordinate system (e.g. G54) instead of G53 machine coordinates.
This change in beneficial as it allows probing programs (such as [smartprobe.ngc](https://github.com/LinuxCNC/linuxcnc/blob/master/nc_files/smartprobe.ngc) and [gridprobe.ngc](https://github.com/LinuxCNC/linuxcnc/blob/master/nc_files/gridprobe.ngc)) which run in a work coordinate system to be run based on specified work coordinate offsets, as standard NC programs typically are.
If compensation instead used G53 coordinates (as it does currently), then the process of running a probing program involves calculating the work coordinate offset and applying it to the probing program so that its results are given in terms of G53 (or otherwise taking probe results in in a work coordinate system and subtracting the values of the work coordinate offsets for X and Y from the respective probed position). This is needlessly complicated and changing compensation to use work coordinate offsets allows for part surfaces to be easily probed and have compensation applied with no extra mental effort on behalf of the operator.
[This post in the LinuxCNC forum shows how the setup process is needlessly complicated when compensation uses machine coordinates.](https://forum.linuxcnc.org/38-general-linuxcnc-questions/41976-adding-auto-leveling-to-linuxcnc?start=10#213111)
In addition, the use of the work coordinate system allows the same surface map to be reused across multiple work coordinate systems and fixtures, if so desired.
It is not expected that this change will affect use for LinuxCNC 3D printing, as if work coordinate offsets are set with a value of X=0 and Y=0, then the behaviour will be the same. That is to say that if value of the G54 work coordinate offsets for X and Y are each 0, then (barring no additional offsets being applied), then G53 X0  = G54 X0, and G53 Y0 = G54 Y0. For 3D printers where a single coordinate system is used with the X and Y work offsets being 0, this change will have no effect, while also improving the experience of users when probing fixtured parts for milling.

_Added recommendation in README to rename HAL signals to avoid conflict with stepconf/pncconf:_
These are the HAL signals currently recommended by the README:
net xpos-cmd 		<= axis.x.pos-cmd			=> compensation.x-pos
net ypos-cmd 		<= axis.y.pos-cmd			=> compensation.y-pos
net zpos-cmd 		<= halui.axis.z.pos-relative		=> compensation.z-pos
However, the signal name N.pos-cmd is already utilised by stepconf and pncconf to connect IN pin stepgen.N.position-cmd to OUT pin joint.N.motor-pos-cmd. As a consequence, the halui.axis.N.pos-relative OUT pin must be connected to a different signal, and for this reason the signals should be renamed:
net eoffset-xpos		<= halui.axis.x.pos-relative		=> compensation.x-pos
net eoffset-ypos 		<= halui.axis.y.pos-relative		=> compensation.y-pos
net eoffset-zpos 		<= halui.axis.z.pos-relative		=> compensation.z-pos
(note this also includes the pin reassignments mentioned above)
[This issue was mentioned by a user in the LinuxCNC forum.](https://forum.linuxcnc.org/38-general-linuxcnc-questions/41976-adding-auto-leveling-to-linuxcnc?start=10#234888)

_Confirm python3 functionality and version requirements in README:_
Compensation.py was confirmed to work in both python2.7 and python3, allowing compensation to be used in LinuxCNC releases that instead utilise python3 (LinuxCNC 2.9 and after). Change README to state this compatibility.
Stated that LinuxCNC 2.8.0 and later are required for External Offsets to function.

_Added reference to smartprobe_compensation.ngc and gridprobe.ngc:_
[This issue was also mentioned by a user in the LinuxCNC forum](https://forum.linuxcnc.org/38-general-linuxcnc-questions/41976-adding-auto-leveling-to-linuxcnc?start=10#203421)

_General addition of details to README:_
Designed to make the repository slightly more accessible to people less experienced with LinuxCNC and to explain the functionality of the program in a little more detail, and to make the installation progress more distinct, explaining some settings (such as OFFSET_AV_RATIO).
(I understand this does make it a little more wordy for experienced users).

**Known issues:**
It appears that specifying high interpolation resolutions (e.g. 0.05mm) in combination with relatively large changes in Z position may result in interpolated points created with values that are slightly smaller than the lowest probed point when using cubic interpolation (as it generates a smooth surface). This may occur between a high point and a point with a value of 0 at the edge of the compensation map.
(e.g. between X0 Z0 and X1 Z20, there may be a minimum of Z-0.1).
I believe that in theory this bug may already be present in the current implementation when using cubic interpolation, as the continuity of this interpolation method means that this issue is likely to occur. However, the low prior resolution masks/aliases this and it is only visible at higher interpolation resolutions.

DauntlessAq committed:
Update readme.md for detail and changes to compensation.py

Improved explanations of component to aid installation and use
Change of explanations for changes made to compensation.py

DauntlessAq committed:
Update readme.md

Rewording

DauntlessAq committed:
Update compensation.py with configurable interpolation resolution

Added configurable interpolation resolution controlled via HAL pin, default value 1.

DauntlessAq committed:
upload modified smartprobe.ngc (smartprobe_compensation.ngc)

Switched to metric from imperial.
Outputs formatted to be used directly by compensation.py.
Changed to named global variables, added annotations.
Added optional slow probe move.
Added 'walking' logic to Y moves to avoid retract to Z safety to wastes time.
Added velocity control to most move types.
Changed to drive the program by final dimensions, with option to round number of rows and/or columns either up or down if the result is not an integer.

DauntlessAq committed:
Move smartprobe_compensation.ngc to compensation directory

DauntlessAq committed:
Update readme.md to include reference to smartprobe_compensation.ngc

zmrdko commented:
Hello, thank you for your effort.
I have going through literally the same issue yesterday: the probe results being in current offset coordinates, etc.

I am trying to add this compensation component to probe_basic UI as a probe tab, to make it more accessible to general public:
![image](https://github.com/user-attachments/assets/89c2d2e3-c369-4c9f-8695-73ac455136da)

DauntlessAq committed:
Update compensation.py

Fixed issue where compensation.clear pin did not remain high until external offset was cleared

DauntlessAq committed:
Update compensation.py

Added hal pin for current eoffset, allows compensation.py to determine if the eoffset has cleared

DauntlessAq committed:
Update readme.md

Added documentation for new compensation.eoffset HAL pin

DauntlessAq committed:
Update compensation.py

Added hal pin for eoffset limited, when clearing compensation now also hold axis.z.eoffset-clear =1 for as long as motion.eoffset-limited = 1 (motion inhibited by soft limit)

DauntlessAq committed:
Update readme.md with extra detail and soft limit logical fix

Added additional detail, including details on usage of External Offsets and a disclaimer.
Details on additional hal pin for handling approaching soft limits. While approaching soft limits is now handled correctly from a logical standpoint, it is not recommended due to the behaviour of External Offsets.

DauntlessAq committed:
Update readme.md

Added statement that movement when enabling/disabling compensation can be avoided by choosing a position where the compensation value is 0.

DauntlessAq committed:
Update compensation.py

Fixed missed closing bracket for eoffset halpin creation
